### PR TITLE
HTTP CONNECT handshaker

### PR DESCRIPTION
**Pull request by markdroth**
The test is not working yet, but I'm not sure why.  There seems to be some weird interaction between the proxy code I've written and the main thread; when I change the code to use tinyproxy instead of running my code in a different thread, the test works fine.

### CODE REVIEW

**Comment by markdroth** [comment]
This is now ready for review.  It depends on #7640.  It fixes #7507.

Note that there's still an open question about how to indicate that we want to use a proxy.  Currently, the code looks for an "http_proxy=" query parameter in the server name URI.  In the future, we can make that do something more appropriate, such as looking for a system-wide proxy config file.

**Comment by markdroth** [comment]
@ctiller, you're welcome to take a look at this as well, but Julien reviewed the rest of the handshaker code, so he's probably the best person to review this.

**Comment by pgrosu** [comment]
They're gonna have fun with 76 files :)

File: src/core/ext/client_config/http_connect_handshaker.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+  // State saved while performing the handshake.
+  grpc_endpoint* endpoint;
+  grpc_channel_args* args;
+  grpc_handshaker_done_cb cb;
+  void* user_data;
+
+  // Objects for processing the HTTP CONNECT request and response.
+  gpr_slice_buffer write_buffer;
+  gpr_slice_buffer* read_buffer;
```
Comment (jboeuf) [inline src/core/ext/client_config/http_connect_handshaker.c]: Looks like read_buffer is not owned by this object. You should probably mark it as such.

File: src/core/ext/client_config/http_connect_handshaker.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  handshaker->server_name = gpr_strdup(server_name);
+  gpr_slice_buffer_init(&handshaker->write_buffer);
+  grpc_closure_init(&handshaker->request_done_closure, on_write_done,
+                    handshaker);
+  grpc_closure_init(&handshaker->response_read_closure, on_read_done,
+                    handshaker);
+  grpc_http_parser_init(&handshaker->http_parser, GRPC_HTTP_RESPONSE,
+                        &handshaker->http_response);
+  gpr_ref_init(&handshaker->refcount, 1);
+  return (grpc_handshaker*)handshaker;
```
Comment (jboeuf) [inline src/core/ext/client_config/http_connect_handshaker.c]: `return &handshaker->base;` avoids a cast (cosmetic).

File: src/core/ext/client_config/http_connect_handshaker.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    grpc_http_parser_destroy(&handshaker->http_parser);
+    grpc_http_response_destroy(&handshaker->http_response);
+    gpr_free(handshaker);
+  }
+}
+
+// Callback invoked when deadline is exceeded.
+static void on_timeout(grpc_exec_ctx* exec_ctx, void* arg, grpc_error* error) {
+  http_connect_handshaker* handshaker = arg;
+  if (error == GRPC_ERROR_NONE)  // Timer fired, rather than being cancelled.
```
Comment (jboeuf) [inline src/core/ext/client_config/http_connect_handshaker.c]: curly brackets for multiline statements please.

**Comment by jboeuf** [comment]
Thanks much for the PR.

the http connect code looks sane to me (a few minor comments). @ctiller should probably have a look at the changes for name resolution and load balancing.

File: src/core/ext/client_config/http_connect_handshaker.c (hunk 1 above)
Comment (markdroth) [inline src/core/ext/client_config/http_connect_handshaker.c]: It's not constructed or destroyed by this object, but it is actually owned by it (at least transitively).  This field is set to the read buffer passed in to the handshaker, which we take ownership of and then pass through to the result callback (which may chain to the next handshaker).  See PR #7640 for details.

I've added a brief comment indicating this.

File: src/core/ext/client_config/http_connect_handshaker.c (hunk 2 above)
Comment (markdroth) [inline src/core/ext/client_config/http_connect_handshaker.c]: Done.

File: src/core/ext/client_config/http_connect_handshaker.c (hunk 3 above)
Comment (markdroth) [inline src/core/ext/client_config/http_connect_handshaker.c]: Done.

**Comment by markdroth** [comment]
It looks like there might still be a problem with the tests.  I'm looking into it.

**Comment by jboeuf** [comment]
Assigning @ctiller for the rest of the PR (name resolution, load balancing...). HTTP CONNECT code LGTM.

**Comment by markdroth** [comment]
The remaining test failures are on windows, so I'm going to need some help from @ctiller and/or @nicolasnoble to debug them.

**Comment by ctiller** [comment]
@nicolasnoble - can you organize some time to help Mark with this?

**Comment by markdroth** [comment]
Thanks to some help from Nico debugging the Windows tests, I suspect that the remaining test failures are not related to this PR, so I think this is ready for a detailed review.

**Comment by jboeuf** [comment]
Do you want me to have another look?

**Comment by markdroth** [comment]
Actually, looks like there may still be a couple of weird Windows test failures.  I'm going to talk with Nico about this again tomorrow.

Julien, you're welcome to look again if you'd like, but I was mainly aiming that comment at Craig. :)

**Comment by jboeuf** [comment]
Works for me :).

**Pull request merged**